homeassistant/components/nut/sensor.py

```diff
@@ -164,65 +57,101 @@
+    """Set up the NUT sensors."""
 
-    _LOGGER.debug("NUT Sensors Available: %s", data.status)
+    config = config_entry.data
+    pynut_data = hass.data[DOMAIN][config_entry.entry_id]
+    data = pynut_data[PYNUT_DATA]
+    status = pynut_data[PYNUT_STATUS]
+    unique_id = pynut_data[PYNUT_UNIQUE_ID]
+    manufacturer = pynut_data[PYNUT_MANUFACTURER]
+    model = pynut_data[PYNUT_MODEL]
+    firmware = pynut_data[PYNUT_FIRMWARE]
 
     entities = []
 
-    for resource in config[CONF_RESOURCES]:
+    name = config[CONF_NAME]
+    if CONF_RESOURCES in config_entry.options:
+        resources = config_entry.options[CONF_RESOURCES]
+    else:
+        resources = config_entry.data[CONF_RESOURCES]
+
+    for resource in resources:
         sensor_type = resource.lower()
 
         # Display status is a special case that falls back to the status value
         # of the UPS instead.
-        if sensor_type in data.status or (
-            sensor_type == KEY_STATUS_DISPLAY and KEY_STATUS in data.status
+        if sensor_type in status or (
+            sensor_type == KEY_STATUS_DISPLAY and KEY_STATUS in status
         ):
-            entities.append(NUTSensor(name, data, sensor_type))
+            entities.append(
+                NUTSensor(
+                    name, data, sensor_type, unique_id, manufacturer, model, firmware
+                )
+            )
         else:
             _LOGGER.warning(
                 "Sensor type: %s does not appear in the NUT status "
                 "output, cannot add",
                 sensor_type,
             )
 
-    try:
-        data.update(no_throttle=True)
-    except data.pynuterror as err:
-        _LOGGER.error(
-            "Failure while testing NUT status retrieval. Cannot continue setup: %s", err
-        )
-        raise PlatformNotReady
-
-    add_entities(entities, True)
+    async_add_entities(entities, True)
 
 
 class NUTSensor(Entity):
     """Representation of a sensor entity for NUT status values."""
 
-    def __init__(self, name, data, sensor_type):
+    def __init__(
+        self, name, data, sensor_type, unique_id, manufacturer, model, firmware
+    ):
         """Initialize the sensor."""
         self._data = data
-        self.type = sensor_type
+        self._type = sensor_type
+        self._manufacturer = manufacturer
+        self._firmware = firmware
+        self._model = model
+        self._device_name = name
         self._name = "{} {}".format(name, SENSOR_TYPES[sensor_type][0])
         self._unit = SENSOR_TYPES[sensor_type][1]
         self._state = None
+        self._unique_id = unique_id
         self._display_state = None
         self._available = False
 
+    @property
+    def device_info(self):
+        """Device info for the ups."""
+        device_info = {
+            "name": self._device_name,
+        }
+        if self._unique_id:
+            device_info["identifiers"] = {(DOMAIN, self._unique_id)}
```

Comment:
Device info will need at least 1 identifier or 1 connection specified or else we can't store it later, as there is nothing to identify the device. So if no unique ID, return None.

Reply:
done